design-proposals/emts-scale-provisioning.md

```diff
@@ -0,0 +1,321 @@
+- Deploy the provisioning service on the local network that support PXE Boot (BIOS/UEFI with DHCP + TFTP) boot and iPXE with HTTPs.  
+- Have a UX to pre-register BareMetal edge nodes using Serial number or UUID or MAC address.
+- Provision different OS profiles to different edge nodes selected based on Serial number or UUID or MAC address.
+- Provision default OS when a device on the LAN boots over PXE and is not pre-registered.
+- Have a ux of  collecting provisioning logs and status of edge nodes.
+
+> Note: It might be possible for EMF-EIM to support provisioning of the EMT-S nodes. supporting this capability as part of MVP depends on any active customer requirements.
+
+## Solution
+
+The solution consists of two major enhancements (modifications) to EMF:
+1) **Support legacy PXE boot to scale EMT-S provisioning** - the EIM will be extended with a local DHCP/TFTP server that helps initiate OS provisioning via legacy PXE boot. 
+2) **Use EIM standalone deployment** - the proposed solution involves deploying a streamlined version of the EIM, referred to as EIM standalone,
+   at the customer's premises, specifically the OXM warehouse. This streamlined EIM will include only the essential components necessary for operating system provisioning. 
+   It is important to note that the EIM standalone will not be a separate branch or fork of the existing EIM system. 
+   The deployment strategy will utilize configuration settings at deployment time within the EMF to ensure that only the required features are activated.
+   The detailed design for the deployment time configuration and profile settings for both EMF and EIM will be addressed in a separate design document.
+
+### Legacy PXE provisioning workflow
+
+The solution relies on the capabilities provided by Tinkerbell SMEE. SMEE provides a minimal DHCP and TFTP server implementations that support PXE boot.
+The DHCP server provides necessary PXE info to machines initiating the PXE boot, while the TFTP server provides the iPXE binary that can be chain-loaded at boot time by PXE.
+The iPXE binary provided by the SMEE's TFTP server is a third-party script, maintained by the Tinkerbell project.
+
+The solution assumes that the local EIM-provided DHCP/TFTP server is only used to initiate boot via legacy PXE (as legacy PXE has limited capabilities, i.e., it doesn't support HTTPS boot),
+but, once chain-loaded to iPXE, the Micro-OS is downloaded and used to drive provisioning.
+In other words, the only difference between the new PXE-based boot and HTTP-based boot is how the OS provisioning is triggered. The subsequent workflow remains the same -
+it leverages Micro-OS, device discovery and Tinkerbell workflow to complete OS provisioning.
+
+**NOTE1**: Customers should provide their own local DHCP server for dynamic IP address assignment.
+
+**NOTE2**: The workflow assumes that the EIM Standalone is deployed locally on customers' premises and ENs have direct connectivity with EIM services.
```

Comment:
shall we say something about the interaction with external (not managed by us) DHCP servers?

Reply:
I'll add a diagram with a sample network topology but we're not really making any changes to network requirements. We had been relying on customer-managed DHCP server for long time already

Reply:
I believe we need to tell the user to change few things in their dhcp server too? Dont we need the next-server option?

Reply:
next-server is provided by SMEE's DHCP server which is configured in proxy DHCP mode

Reply:
If we enforce any change to DHCP servers, this design proposal wouldn't make sense to me

Reply:
ok this is what I found to make them coexist:
- A PXE client sends a DHCP request (broadcast).
- The external DHCP server responds with: An IP address, The next-server (Boots server IP), The filename (iPXE binary).
- Tinkerbell SMEE (in proxy DHCP mode) responds with PXE-specific options (e.g., boot filename and TFTP/HTTP server details).
- The PXE client downloads the iPXE binary and executes the boot script provided by Boots.

Reply:
let me do some more experiments to confirm once I have my target on-prem environment running

Reply:
after experiments, I've updated the proposal. Also added network diagram to clarify network reqs